lib/SILOptimizer/Transforms/CopyForwarding.cpp

```diff
@@ -1324,10 +1325,244 @@ class CopyForwardingPass : public SILFunctionTransform
       invalidateAnalysis(SILAnalysis::InvalidationKind::CallsAndInstructions);
       }
   }
+};
+
+/// Temporary RValue Optimization
+///
+/// Peephole optimization to eliminate short-lived immutable temporary copies.
+/// This handles a common pattern generated by SILGen where temporary RValues
+/// are emitted as copies...
+///
+///   %temp = alloc_stack $T
+///   copy_addr %src to [initialization] %temp : $*T
+///   // no writes to %src and %temp
+///   destroy_addr %temp : $*T
+///   dealloc_stack %temp : $*T
+///
+/// This differs from the copy forwarding algorithm because it handles
+/// copy source and dest lifetimes that are unavoidably overlappying. Instead,
+/// it finds cases in which it is easy to determine that the source is
+/// unmodified during the copy destination's lifetime. Thus, the destination can
+/// be viewed as a short-lived "rvalue".
+///
+/// The optimization is identified in two parts.
+///
+/// 1. Read-Only Temp Copy: The copy's destination is an AllocStack that is
+/// only written by the initializing copy.
+///
+/// 2. No instructions between the copy and the destroy of its destination
+/// writes to the source. This is sufficient to prove the copy is unnecesary.
+class TempRValueOptPass : public SILFunctionTransform
+{
+  AliasAnalysis *AA = nullptr;
+
+  static bool collectLoads(SILInstruction *CurrentInst, SILInstruction *addr,
+                           llvm::SmallPtrSetImpl<SILInstruction *> &loadInsts);
 
+  bool checkNoSourceModification(CopyAddrInst *copyInst,
+                       const llvm::SmallPtrSetImpl<SILInstruction *> &useInsts);
+
+  bool tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst);
+
+  void run() override;
 };
+
+/// The main entry point of the pass.
+void TempRValueOptPass::run() {
+  DEBUG(llvm::dbgs() << "Copy Peephole in Func " << getFunction()->getName()
+        << "\n");
+
+  AA = PM->getAnalysis<AliasAnalysis>();
+  bool Changed = false;
+
+  // Find all copy_addr instructions.
+  for (auto &BB : *getFunction()) {
+    auto II = BB.begin();
+    while (II != BB.end()) {
+      CopyAddrInst *CopyInst = dyn_cast<CopyAddrInst>(&*II);
+
+      if (CopyInst) {
+        // In case of success, this may delete instructions, but not the
+        // CopyInst itself.
+        Changed |= tryOptimizeCopyIntoTemp(CopyInst);
+      }
+
+      // Increment the instruction iterator now in case the CopyInst is deleted.
+      ++II;
```

Comment:
Because it would crash if the next instruction after the copy_addr is one that is deleted in the optimization, e.g. destroy_addr.
Very unlikely but it could happen.